### PR TITLE
Implementation of Ratcliff-Obershelp algorithm

### PR DESCRIPTION
**Pull request by denmase**
Ported from .Net code by Ligi https://github.com/dxpux (as a patch for fuzzystring)
I've edited the README.md as well, but since I'm not an expert to it, so rework probably is needed.

### CODE REVIEW

File: src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        if (target == null) {
+            throw new NullPointerException("target must not be null");
+        }
+
+        if (source.equals(target)) {
+            return 1;
+        }
+
+        List<String> matches; // = new ArrayList<>();
+        matches = getMatchQueue(source, target);
```
Comment (paulirwin) [inline src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java:70]: These lines can be combined, and the comment removed.

File: src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        int sumOfMatches = 0;
+        Iterator it;
+        it = matches.iterator();
```
Comment (paulirwin) [inline src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java:73]: These lines can be combined

File: src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+        // Display element by element using Iterator
```
Comment (paulirwin) [inline src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java:75]: Probably remove this comment

File: src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        List<String> matches; // = new ArrayList<>();
+        matches = getMatchQueue(source, target);
+        int sumOfMatches = 0;
+        Iterator it;
+        it = matches.iterator();
+
+        // Display element by element using Iterator 
+        while (it.hasNext()) {
+            String element = it.next().toString();
+            //System.out.println(element);
```
Comment (paulirwin) [inline src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java:78]: Remove comment

File: src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            sumOfMatches += element.length();
+        }
+        return 2.0d * sumOfMatches / (source.length() + target.length());
```
Comment (paulirwin) [inline src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java:81]: This is me being nitpicky, but can we throw a blank line before this `return`? :smile:

Comment (paulirwin) [reply src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java:81]: Also for the other if/return blocks in this file

Comment (denmase) [reply src/main/java/info/debatty/java/stringsimilarity/RatcliffObershelp.java:81]: Sure, it's coding standard, it's not nitpicking at all. I'll try to revert back tomorrow. Thank you.

**Comment by denmase** [comment]
Since there were too many changes, although most of them are cosmetics changes, I'll close the pull request and submit another one.

**Comment by tdebatty** [comment]
Hi! @paulirwin and @denmase ! Thank you for your contribution!
Fully agree with the process here.
As soon as the code is cleaned and there is unit testing, I'll be happy to merge and release a new version...

**Comment by denmase** [comment]
@tdebatty Thank you. Yes, just now I've added unit testing as well. I have just found out as well that I can *actually* do style-check using Maven, sorry for being such noob. I have question though, when it passes the style-check, is it deemed to satisfy the required level of coding standard?

**Comment by tdebatty** [comment]
@denmase yes, this is 'good enough' for me ;-)

**Pull request closed without merging**